src/core/mount.c

```diff
@@ -2458,6 +2461,58 @@ static const char* const mount_result_table[_MOUNT_RESULT_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(mount_result, MountResult);
 
+static bool mount_can_freeze(Unit *u) {
+        Mount *m = MOUNT(u);
+        const char *fstype;
+
+        assert(u);
+        assert(m);
+
+        fstype = mount_get_fstype(m);
+        if (!fstype)
+                return false;
+
+        return fstype_can_freeze_thaw(fstype);
+}
+
+static int mount_freezer_action(Unit *u, FreezerAction action) {
+        _cleanup_close_ int fd = -EBADF;
+        Mount *m = MOUNT(u);
+        FreezerState next, target;
+        int r;
+
+        assert(u);
+        assert(m);
+        assert(IN_SET(action, FREEZER_FREEZE, FREEZER_PARENT_FREEZE,
+                              FREEZER_THAW, FREEZER_PARENT_THAW));
+
+        unit_next_freezer_state(u, action, &next, &target);
+        next = freezer_state_finish(next); /* We're completely sync */
+
+        log_unit_debug(u, "Mount freezer state was %s, now %s.",
+                       freezer_state_to_string(u->freezer_state),
+                       freezer_state_to_string(next));
+
+        fd = open(m->where, O_RDONLY|O_NOCTTY);
```

Comment:
O_CLOEXEC

Reply:
O_DIRECTORY

Reply:
> O_DIRECTORY

We may be bind-mounting over a file, right? So we can't do O_DIRECTORY

Reply:
Though I suppose you can't actually freeze bind mounts, so you're right. O_DIRECTORY should be there